stablehlo/compatibility/dialect/VersionedStablehloOps.td

```diff
@@ -0,0 +1,976 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+   Copyright 2022 The StableHLO Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef STABLEHLO_COMPATIBILITY_VERSIONEDHLO_OPS
+#define STABLEHLO_COMPATIBILITY_VERSIONEDHLO_OPS
+
+include "stablehlo/compatibility/dialect/VersionedStablehloEnums.td"
+include "stablehlo/compatibility/dialect/VersionedStablehloAttrs.td"
+
+include "mlir/Dialect/Shape/IR/ShapeBase.td"
+include "mlir/IR/OpBase.td"
+
+//===----------------------------------------------------------------------===//
+// Dialect and Ops
+//===----------------------------------------------------------------------===//
+
+def VersionInterface : OpInterface<"VersionInterface"> {
+  let methods = [
+    InterfaceMethod<
+      "Description...", "mlir::versionedhlo::VersionNumber", "getMinVersion">,
+    InterfaceMethod<
+      "Description...", "mlir::versionedhlo::VersionNumber", "getMaxVersion">,
+  ];
+}
+
+// Only use this base class if necessary.
+// - ReturnOp needs a trait for Terminator.
+// - ReduceOp/ReduceWindowOp/ScatterOp need a trait since they have
+//   multiple variadic arguments.
+class VersionedHLO_OpWithTraits<string mnemonic, list<Trait> traits, string minVersion = "0.0.0", string maxVersion = "current"> :
+    Op<VersionedHLO_Dialect, mnemonic, 
+      [DeclareOpInterfaceMethods<VersionInterface>] # traits> {
+  let extraClassDefinition = [{
+    mlir::versionedhlo::VersionNumber $cppClass::getMinVersion() {
+      return *mlir::versionedhlo::VersionNumber::get("}] #  minVersion # [{");
+    }
+    mlir::versionedhlo::VersionNumber $cppClass::getMaxVersion() {
+      return *mlir::versionedhlo::VersionNumber::get("}] #  maxVersion # [{");
+    }
+  }];
+}
+
+class VersionedHLO_Op<string mnemonic, string minVersion = "0.0.0", string maxVersion = "current"> :
+    VersionedHLO_OpWithTraits<mnemonic, [], minVersion, maxVersion>;
```

Comment:
My thoughts:
1. It is more difficult to misuse this way, must deliberately choose this class for traits.
2. There is some slight code-ugliness if we don't in that every op with a non-default version will need to specify empty trait list:

```
def VHLO_CustomCallOpV1 : VHLO_Op<"custom_call", [], "0.0.0", "0.0.0"> 
def VHLO_CustomCallOpV2 : VHLO_Op<"custom_call", [], "0.1.0"> 
```

The alternative is to make the trait list the last argument, in which case only ops that need traits have this ugliness, need to explicitly specify the default values to specify the traits:

```
def VHLO_ReturnOpV1 : VHLO_Op<"return", "0.0.0", "current", [Terminator]>
def VHLO_ReduceOpV1 : VHLO_Op<"reduce", "0.0.0", "current", [SameVariadicOperandSize]>
def VHLO_ScatterOpV1 : VHLO_Op<"scatter", "0.0.0", "current", [SameVariadicOperandSize]>
def VHLO_ReduceWindowOpV1 : VHLO_Op<"reduce_window", "0.0.0", "current", [SameVariadicOperandSize]> 
```

I'll go with the latter route for now since there are only those 4 ops that need traits currently. Removed `OpWithTraits` for now.